Fix data formatters for Swift Set and Dictionary types

lorentey opened the pull request:
This PR updates lldb's Swift Set and Dictionary data formatters, removing obsolete paths, simplifying code structure and fixing edge cases.

rdar://problem/43045289

Partial list of fixes:
- Fixes summary formatters for Swift sets -- they should be summaried as `<n> values`, like arrays.
    ```
    $ lldb --repl
    1> import Foundation

    // Before: ❌
    2> [1, 2] as Set<Int>
    $R0: Set<Int> = {  // Note: no summary
      [0] = 1
      [1] = 2
    }
    // After: 👍
    2> [1, 2] as Set<Int>
    $R0: Set<Int> = 2 values {
       [0] = 1
       [1] = 2
     }
    ```
- Empty sets and dictionaries using the type-punned singleton storage instance are now formatted correctly. This also works when the storage instance is verbatim bridged to Objective-C.
    ```
    // Before: ❌
    2> Set<Int>()
    $R0: Set<Int> = {} 
    3> $R0 as NSSet
    $R1: _RawNativeSetStorage = {
      Swift._SwiftNativeNSSet = {}
      bucketCount = 1
      count = 0
      ... more gobbledygook ...
    }

    // After: 👍
    2> Set<Int>()
    $R0: Set<Int> = 0 values
    3> $R0 as NSSet
    $R1: _RawNativeSetStorage = 0 values
    ```

- Swift sets and dictionaries that are bridged to Objective-C now use the proper formatter, rather than a raw data dump:
    ```
    // Before: ❌
    2> [1, 2] as Set<Int> as NSSet
    $R0: _SwiftDeferredNSSet<Int> = {
       Swift._SwiftNativeNSSet = {}
       _heapStorageBridged_DoNotUse = nil
       nativeBuffer = {
         _storage = {
           ... more gobbledygook ...
         }
       }
    }
    3> [NSNumber(value: 1), NSNumber(value: 2)] as Set<NSObject> as NSSet
    $R1: _HashableTypedNativeSetStorage<NSObject> = {
      Swift._TypedNativeSetStorage = {
        Swift._RawNativeSetStorage = {
          Swift._SwiftNativeNSSet = {}
          bucketCount = 4
           ... more gobbledygook ...
        }
      }
    }

    // After: 👍
    2> [1, 2] as Set<Int> as NSSet
    $R0: _SwiftDeferredNSSet<Int> = 2 values {
      [0] = 1
      [1] = 2
    }
    3> [NSNumber(value: 1), NSNumber(value: 2)] as Set<NSObject> as NSSet
    $R1: _HashableTypedNativeSetStorage<NSObject> = 2 values {
      [0] = Int64(2)
      [1] = Int64(1)
    }
    ```
- Cocoa NSSet and NSDictionary instances that are verbatim-bridged to Swift are now formatted correctly:
    ```
    // Before: ❌
    2> NSSet(array: [NSNumber(value: 1), NSNumber(value: 2)]) as! Set<NSObject>
    $R0: Set<NSObject> = {}

    // After: 👍
    2> NSSet(array: [NSNumber(value: 1), NSNumber(value: 2)]) as! Set<NSObject>
    $R0: Set<NSObject> = 2 values {
      [0] = Int64(1)
      [1] = Int64(2)
    }
    ```

lorentey commented:
@swift-ci please test

dcci commented on source/Plugins/Language/Swift/SwiftDictionary.cpp:
```diff
@@ -26,95 +26,48 @@ using namespace lldb_private;
+
+  // Native storage class
+  m_nativeStorage_mangledPrefix =
+    ConstString(SwiftLanguageRuntime::GetCurrentMangledName("_TtGCs37_HashableTypedNativeDictionaryStorage"));
```
You can probably use the new mangled name here.

lorentey replied:
We apparently still use this mangling for the Objective-C runtime names of NSObject subclasses defined in Swift.

The name is incomplete, so while I cargo-culted GetCurrentMangledName over from the old code, it actually has no effect.

dcci commented on source/Plugins/Language/ObjC/NSArray.cpp:
```diff
@@ -35,17 +35,16 @@ using namespace lldb_private::formatters;
 namespace lldb_private {
 namespace formatters {
-std::map<ConstString, CXXFunctionSummaryFormat::Callback> &
+NSArray_Additionals::SummaryFormatters &
```
This looks like it's code in `ObjC/`. Is this Swift specific? If so, maybe we should move it to a `Swift/` specific directory to reduce the amount of conflicts with upstream. Otherwise, if this might be useful for `NSArray` et similia also when running in `ObjC` language mode, this patch (or at least, part of it) should live upstream.

lorentey replied:
These are hooks that Swift inserted so that it can hook into the Objective-C mode's formatter dispatch for Swift's implementations of Foundation collections. ObjC could use these to register its own formatters, too, but it doesn't really need it -- the if ladders it currently has work fine.

The new ObjC formatter (9b0178f172c284a507559d022374cead013fb152) and related NSSet formatter fixes (790e5f6cb1a954b0c778a54568932f6f198d11c6) are entirely Objective-C specific and should definitely live upstream!

lorentey replied:
Ah, wait, I see the hooks are also in the upstream repo. You're right, the Objective-C parts of 66cbe5880522688825c93b91c51470c0aea2149a should also get merged there!

dcci replied:
I'd very much recommend to just submit the patch upstream (I'll be happy to review it!) and then cherry-pick here (at least, this is the flow we've been following for a while).

dcci reviewed (commented):
I had a quick glance at this patch and the swift bits look OK. I have a more general question about the structure of this patch before going in the details, see comment(s) inline.

lorentey committed:
[Swift] Add REPL torture tests for Set and Dictionary edge cases

lorentey committed:
[Swift] Remove ObjC formatter hooks for Set & Dictionary

AFAICT they are unnecessary -- simply registering Swift formatters for the ObjC class names of storage/bridging classes works just as well.

lorentey committed:
[Swift] Update Set and Dictionary class names

- Replace obsolete (Swift 3 and below) storage class names with current names.
- Add support for formatting type-punned empty sets/dictionaries.

lorentey committed:
[Swift] Refactor Set, Dictionary handlers

- Collect Set/Dictionary information into a new class, HashedCollectionConfig. Use it as a mixin instead of defining separate handler subclasses.
- Move all Set/Dictionary type name patterns into HashedCollectionConfig.
- Move handler creation into HashedCollectionConfig.
- Fix detection of native storage classes.
- Fix detection of the type-punned empty storage singletons.
- Fix detection of _SwiftDeferredNSSet.

Additionally, perform the following renames:

HashedContainerSyntheticFrontEnd => HashedCollectionSyntheticChildrenFrontEnd SwiftHashedContainerBufferHandler => HashedStorageHandler SwiftHashedContainerEmptyBufferHandler => EmptyHashedStorageHandler SwiftHashedContainerNativeBufferHandler => NativeHashedStorageHandler SwiftHashedContainerSyntheticFrontEndBufferHandler => CocoaHashedStorageHandler

Move the last three classes out of SwiftHashedContainer.h.

lorentey committed:
[Swift] Review native storage handler

- Radically simplify native storage layout discovery.
- Remove support for obsolete layouts.
- Fix Swift handler for bridged Cocoa objects.
- Stronger validation for the bucket count
- Rename m_capacity to m_bucketCount for consistency with Swift sources

lorentey committed:
[Swift] Also register formatters for ObjC mangled class names

lorentey committed:
[Swift] Ensure input is a Swift object before looking at it

When a Swift hashed storage class appears in Objective-C context, the supplied ValueObject is an Objective-C pointer type. Get the address it contains and convert it to Swift.AnyObject before looking at its contents.

lorentey committed:
[Swift] Use the dynamic type of the storage object to resolve key/value types

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey commented:
@dcci I rebased this & restructured the code, removing the parts that touch upstream code. I isolated the NSSet formatter fixes to a separate branch; I'll submit them upstream next week.

It turns out we don't need to hook into Foundation's formatter dispatch; AFAICT, merely registering the Swift formatters with the Objective-C class names works fine, as long as the formatters are prepared to take ObjC pointers as well as Swift values. (The value conversion in `HashedCollectionConfig::SwiftObjectAtAddress` is the messy bit.)

lorentey committed:
Revert "[Swift] Use the dynamic type of the storage object to resolve key/value types"

This reverts commit 2476a02c903d23b8a1757abfcb9c3fe48be12773.

lorentey commented:
```
FAIL: test_dictionary_nsobject_any_object_dwarf (TestSwiftDictionaryNSObjectAnyObject.TestDictionaryNSObjectAnyObject)
AssertionError: False is not True : 'frame variable -d run -- d2' returns expected result, 
got '([NSObject : AnyObject]) d2 = 0 key/value pairs {}'
```

lldb/packages/Python/lldbsuite/test/lang/swift/variables/dict_nsobj_anyobj/main.swift:
```
  var d1: Dictionary<Int,Int> = [1:1,2:2,3:3,4:4]
  var d2: Dictionary<NSObject,AnyObject> = [1 as NSNumber : 1 as NSNumber,
                                            2 as NSNumber : 2 as NSNumber,
                                            3 as NSNumber : 3 as NSNumber,
                                            4 as NSNumber : 4 as NSNumber]
```

Hm. d1 seems to work fine, but we fail to get the dynamic type of d2's storage reference -- it somehow resolves to its static type, `Swift._RawNativeDictionaryStorage`. This base class is used for the empty singleton instance, so we get on that path, and print d2 as having 0 elements. Weird; I don't know why the dynamic type fails to resolve in this particular case.

Reverting 2476a02c903d23b8a1757abfcb9c3fe48be12773 fixes the issue, but it removes dynamic isa checks for Dictionary variables -- so uninitialized values may cause confusion more often. (This is not a regression; lldb wasn't verifying the storage isa before this PR, either.) I think this is fine for now: the PR already improves storage layout verification elsewhere.

lorentey commented:
@swift-ci test

lorentey committed:
[Swift] Prepare for upcoming round of renamings in stdlib

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci please test

dcci commented:
This looks great, thanks (sorry for the delay in replying, I was OOO last week)
